barretenberg/cpp/src/barretenberg/commitment_schemes/shplonk/shplemini.hpp

```diff
@@ -439,5 +466,70 @@ template <typename Curve> class ShpleminiVerifier_ {
             commitments.emplace_back(std::move(fold_commitments[j]));
         }
     }
+
+    /**
+     * @brief Add the opening data corresponding to Libra masking univariates to the batched opening claim
+     *
+     * @details After verifying ZK Sumcheck, the verifier has to validate the claims about the evaluations of Libra
+     * univariates used to mask Sumcheck round univariates. To minimize the overhead of such openings, we continue the
+     * Shplonk batching started in Gemini, i.e. we add new claims multiplied by a suitable power of the Shplonk batching
+     * challenge and re-use the evaluation challenge sampled to prove the evaluations of Gemini polynomials.
+     *
+     * @param commitments
+     * @param scalars
+     * @param libra_univariate_commitments
+     * @param libra_univariate_evaluations
+     * @param multivariate_challenge
+     * @param shplonk_batching_challenge
+     * @param shplonk_evaluation_challenge
+     */
+    static void add_zk_data(std::vector<Commitment>& commitments,
+                            std::vector<Fr>& scalars,
+                            RefSpan<Commitment> libra_univariate_commitments,
+                            const std::vector<Fr>& libra_univariate_evaluations,
+                            const std::vector<Fr>& multivariate_challenge,
+                            const Fr& shplonk_batching_challenge,
+                            const Fr& shplonk_evaluation_challenge)
+
+    {
+        // compute current power of Shplonk batching challenge taking into account the const proof size
+        Fr shplonk_challenge_power = Fr{ 1 };
+        for (size_t j = 0; j < CONST_PROOF_SIZE_LOG_N + 2; ++j) {
+            shplonk_challenge_power *= shplonk_batching_challenge;
+        }
+
+        // needed to keep track of the constant term contribution
+        const size_t idx_of_constant_term = commitments.size() - 1;
+
+        // compute shplonk denominators and batch invert them
+        std::vector<Fr> denominators;
+        size_t num_libra_univariates = libra_univariate_commitments.size();
+
+        // compute Shplonk denominators and invert them
+        for (size_t idx = 0; idx < num_libra_univariates; idx++) {
```

Comment:
Is the number of libra univariates different from the size of the multivariate challenge? If not it might be better to use a range loop here on `multivariate_challenge`